FX-8350 CPU P-state frequencies are doubled, but CPU-NB frequencies are correctly listed?

Hi There,

Loving `amdctl` and am using it on my old PhII and FX-8350 boxes.

I get this output when listing the states:

```
$ sudo ./amdctl -g -c0
Voltage ID encodings: SVI (serial)
Detected CPU model 2h, from family 15h with 8 CPU cores.

Core 0 | P-State Limits (non-turbo): Highest: 1 ; Lowest 5 | Current P-State: 4
 Pstate Status CpuFid CpuDid CpuVid CpuMult CpuFreq CpuVolt IddVal IddDiv CpuCurr CpuPower
      0      1     26      0     10  21.00x 8400MHz  1425mV    134     10  13.40A   19.09W
      1      1     25      0     13  20.50x 8200MHz  1388mV    134     10  13.40A   18.59W
      2      1     24      0     18  20.00x 8000MHz  1325mV    119     10  11.90A   15.77W
      3      1     18      0     27  17.00x 6800MHz  1212mV     95     10   9.50A   11.52W
      4      1     12      0     35  14.00x 5600MHz  1112mV     75     10   7.50A    8.34W
      5      1      5      0     44  10.50x 4200MHz  1000mV     54     10   5.40A    5.40W
current      1      5      0     44  10.50x 4200MHz  1000mV
Northbridge:
P-State 0: 28 (vid),  1200mV, 2400MHz
P-State 1: 31 (vid),  1162mV, 2200MHz
```

The CPU-NB frequencies are correct, but the CPU P-state frequencies are doubled?

I took a short look at the code, but it seems that the CPU-NB and the CPU P-state frequencies require different interpretations? I'm happy to try out different approaches per your suggestions and then creating a PR once I get it working correctly FWIW.

Tried switching out the existing code with the edited function you just pasted and this is what I got (this is a different box than the first one I used -- I suspect this box has a BIOS bug with the CPU-NB P-State settings):

```
ermo@solbox:~/repos/amdctl ⑂master*
$ git diff
diff --git a/amdctl.c b/amdctl.c
index 4b0182c..08a2bbb 100644
--- a/amdctl.c
+++ b/amdctl.c
@@ -626,11 +626,10 @@ float getCpuMultiplier(const int CpuFid, const int CpuDid) {
 int getClockSpeed(const int CpuFid, const int CpuDid) {
        switch (cpuFamily) {
                case AMD10H:
+               case AMD11H:
                case AMD15H:
                case AMD16H:
-                       return ((REFCLK * (CpuFid + 0x10)) >> CpuDid);
-               case AMD11H:
-                       return ((REFCLK * (CpuFid + 0x08)) >> CpuDid);
+                       return (REFCLK * (CpuFid + (cpuFamily == AMD11H ? 0x08 : 0x10)) / (int) pow(2.0, (float) CpuDid));
                case AMD12H:
                        return (int) (REFCLK * (CpuFid + 0x10) / getDiv(CpuDid));
                case AMD17H:
ermo@solbox:~/repos/amdctl ⑂master*
$ make
gcc -Wall -pedantic -Wextra -std=c99 -O2 -lm    amdctl.c   -o amdctl
amdctl.c:29:29: warning: Incompatible kernel! MSR access deprecated by your Linux kernel! To use this program, set the kernel parameter msr.allow_writes=on or set /sys/module/msr/parameters/allow_writes to on at runtime.
   29 |         #pragma GCC warning "Incompatible kernel! MSR access deprecated by your Linux kernel! To use this program, set the kernel parameter msr.allow_writes=on or set /sys/module/msr/parameters/allow_writes to on at runtime."
      |                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ermo@solbox:~/repos/amdctl ⑂master*
$ sudo ./amdctl -g -c0
Voltage ID encodings: SVI (serial)
Detected CPU model 2h, from family 15h with 8 CPU cores.

Core 0 | P-State Limits (non-turbo): Highest: 1 ; Lowest 5 | Current P-State: 1
 Pstate Status CpuFid CpuDid CpuVid CpuMult CpuFreq CpuVolt IddVal IddDiv CpuCurr CpuPower
      0      1     28      0     10  22.00x 8800MHz  1425mV    134     10  13.40A   19.09W
      1      1     26      0     12  21.00x 8400MHz  1400mV    134     10  13.40A   18.76W
      2      1     24      0     16  20.00x 8000MHz  1350mV    119     10  11.90A   16.06W
      3      1     18      0     24  17.00x 6800MHz  1250mV     95     10   9.50A   11.88W
      4      1     12      0     32  14.00x 5600MHz  1150mV     75     10   7.50A    8.62W
      5      1      5      0     42  10.50x 4200MHz  1025mV     54     10   5.40A    5.54W
current      1     12      1     12   7.00x 2800MHz  1400mV
Northbridge:
P-State 0: 29 (vid),  1188mV, 4400MHz
P-State 1: 0 (vid),  1550mV, 1600MHz
ermo@solbox:~/repos/amdctl ⑂master*
$ gcc --version
gcc (Solus) 11.2.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

Looks like it now doesn't divide by 2 like it's supposed to -- neither in the CPU P-states, nor in the NB P-states?

It seems like it's supposed to be 2^(CpuDid+1), such that it's either dividing by 2 (when CpuDid = 0) or 4 (when CpuDid = 1)?

It looks like this if I try my hand at it:

```
ermo@solbox:~/repos/amdctl ⑂master*
$ git diff
diff --git a/amdctl.c b/amdctl.c
index 4b0182c..bc42bf9 100644
--- a/amdctl.c
+++ b/amdctl.c
@@ -626,11 +626,10 @@ float getCpuMultiplier(const int CpuFid, const int CpuDid) {
 int getClockSpeed(const int CpuFid, const int CpuDid) {
        switch (cpuFamily) {
                case AMD10H:
+               case AMD11H:
                case AMD15H:
                case AMD16H:
-                       return ((REFCLK * (CpuFid + 0x10)) >> CpuDid);
-               case AMD11H:
-                       return ((REFCLK * (CpuFid + 0x08)) >> CpuDid);
+                       return (int) (REFCLK * (CpuFid + (cpuFamily == AMD11H ? 0x08 : 0x10)) / pow(2.0, (float) CpuDid+1));
                case AMD12H:
                        return (int) (REFCLK * (CpuFid + 0x10) / getDiv(CpuDid));
                case AMD17H:
ermo@solbox:~/repos/amdctl ⑂master*
$ make
gcc -Wall -pedantic -Wextra -std=c99 -O2 -lm    amdctl.c   -o amdctl
amdctl.c:29:29: warning: Incompatible kernel! MSR access deprecated by your Linux kernel! To use this program, set the kernel parameter msr.allow_writes=on or set /sys/module/msr/parameters/allow_writes to on at runtime.
   29 |         #pragma GCC warning "Incompatible kernel! MSR access deprecated by your Linux kernel! To use this program, set the kernel parameter msr.allow_writes=on or set /sys/module/msr/parameters/allow_writes to on at runtime."
      |                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ermo@solbox:~/repos/amdctl ⑂master*
$ sudo ./amdctl -g -c0
Voltage ID encodings: SVI (serial)
Detected CPU model 2h, from family 15h with 8 CPU cores.

Core 0 | P-State Limits (non-turbo): Highest: 1 ; Lowest 5 | Current P-State: 1
 Pstate Status CpuFid CpuDid CpuVid CpuMult CpuFreq CpuVolt IddVal IddDiv CpuCurr CpuPower
      0      1     28      0     10  22.00x 4400MHz  1425mV    134     10  13.40A   19.09W
      1      1     26      0     12  21.00x 4200MHz  1400mV    134     10  13.40A   18.76W
      2      1     24      0     16  20.00x 4000MHz  1350mV    119     10  11.90A   16.06W
      3      1     18      0     24  17.00x 3400MHz  1250mV     95     10   9.50A   11.88W
      4      1     12      0     32  14.00x 2800MHz  1150mV     75     10   7.50A    8.62W
      5      1      5      0     42  10.50x 2100MHz  1025mV     54     10   5.40A    5.54W
current      1     26      0     12  21.00x 4200MHz  1400mV
Northbridge:
P-State 0: 29 (vid),  1188mV, 4400MHz
P-State 1: 0 (vid),  1550mV, 1600MHz
ermo@solbox:~/repos/amdctl ⑂master*
$ 
```
... but then the CPU-NB frequency shown is wrong.

Thoughts?

--- a/amdctl.c
+++ b/amdctl.c
@@ -77,7 +77,7 @@ void northBridge(const int);
 #define VID_DIVIDOR2 12.5
 #define VID_DIVIDOR3 6.25
 
-static int  REFCLK          = 100;
+static const int REFCLK     = 100;  // this is considered a read-only invariant!
 static char *NB_VID_BITS    = "31:25";
 static char *CPU_DID_BITS   = "8:6";
 static char *CPU_FID_BITS   = "5:0";
@@ -209,7 +209,7 @@ int main(int argc, char **argv) {
 
 	if (!quiet) {
 		printf("Voltage ID encodings: %s\n", (pvi ? "PVI (parallel)" : "SVI (serial)"));
-		printf("Detected CPU model %xh, from family %xh with %d CPU cores.\n", cpuModel, cpuFamily, cores);
+		printf("Detected CPU model %xh, from family %xh with %d CPU cores (REFCLK = %dMHz).\n", cpuModel, cpuFamily, cores, REFCLK);
 		if (nv > -1 || cv > -1 || fid > -1 || did > -1) {
 			printf("%s\n", (testMode ? "Preview mode On - No P-State values will be changed."
 									 : "PREVIEW MODE OFF - P-STATES WILL BE CHANGED!"));
@@ -350,17 +350,26 @@ void northBridge(const int nvid) {
 			} else {
 				return;
 			}
+			// We need to be able to display the REFCLK used for the calculations
+			int _refclk = REFCLK;
+			if (cpuModel >= 0x00 && cpuModel <= 0x0f)
+			{
+				_refclk = 2 * REFCLK;
+			}
 			for (int nbpstate = 0; nbpstate < nbpstates; nbpstate++) {
 				getAddr("18.5", addresses[0][nbpstate]);
 				nbvid = ((getDec("16:10") + (getDec("21:21") << 7)));
 				nbfid = getDec("7:7");
 				nbdid = getDec("6:1");
 				printf(
-						"P-State %d: %d (vid), %5.0fmV, %1.0fMHz\n",
+						"P-State %d: %d (vid), %d (fid), %d (did), %5.0fmV, %dMHz (REFCLK = %dMHz)\n",
 						nbpstate,
 						nbvid,
+						nbfid,
+						nbdid,
 						vidTomV(nbvid),
-						(((cpuModel >= 0x00 && cpuModel <= 0x0f) ? REFCLK * 2 : REFCLK) * (nbdid + 0x4) / pow(2, nbfid))
+						(_refclk * (nbdid + 0x4) >> nbfid),
+						_refclk
 				);
 			}
 			break;
@@ -431,9 +440,6 @@ void checkFamily() {
 			if (cpuModel > 0x0f) {
 				NB_VID_BITS = "31:24";
 			}
-			if (cpuModel >= 0x00 && cpuModel <= 0x0f) {
-				REFCLK = 200;
-			}
 			break;
 		case AMD16H:
 			NB_VID_BITS = "31:24";
@@ -449,7 +455,7 @@ void checkFamily() {
 		case AMD13H:
 		case AMD14H: // Disabled due to differences in cpu vid / did / fid
 		default:
-			error("Your CPU family is unsupported by amdctl.");
+			error("Your CPU family is not supported by amdctl.");
 	}
 }
 

